File: Python/rule_parser.py
"""
Parses YAML rule files into Rule objects.

YAML format:
  rules:
    - id: my_rule
      description: "..."
      given:
        - create $obj: ClassName($arg)
        - call: $obj.method($arg)
        - mock: $mock.some_func returns value
      assert:
        - IsNone: $obj.attr
        - AreEqual: [$a, $b]
        - Raises: ValueError
"""

import re
import yaml
from models import (
    Rule, AssertClause, GivenClause,
    CreateClause, CallClause, AssignClause, MockClause,
)


def parse(yaml_content: str) -> list[Rule]:
    doc = yaml.safe_load(yaml_content)
    return [_parse_rule(yr) for yr in doc["rules"]]


def _parse_rule(yr: dict) -> Rule:
    rule = Rule(id=yr["id"], parent=yr.get("parent"), description=yr.get("description", ""))
    for given_map in yr.get("given") or []:
        rule.given.append(_parse_given_clause(given_map))
    for assert_map in yr.get("assert") or []:
        rule.asserts.append(_parse_assert_clause(assert_map))
    return rule


# ── Given clause parsing ─────────────────────────────

def _parse_given_clause(mapping: dict) -> GivenClause:
    key, value = next(iter(mapping.items()))
    val = str(value) if value is not None else ""

    # Check for "not " prefix → negated clause
    negated = False
    if key.lower().startswith("not "):
        negated = True
        key = key[4:]  # strip "not "

    clause = _parse_given_clause_core(key, val)
    clause.negated = negated
    return clause


def _parse_given_clause_core(key: str, val: str) -> GivenClause:
    # create $var: TypePattern($arg1, $arg2)
    m = re.match(r"^create\s+(\$\w+)$", key)
    if m:
        return _parse_create_clause(m.group(1), val)

    # $result = call: ...
    m = re.match(r"^(\$\w+)\s*=\s*call$", key)
    if m:
        return _parse_call_clause(val, result_bind_var=m.group(1))

    # call: ...
    if key == "call":
        return _parse_call_clause(val)

    # assign $target = $source: CastType
    m = re.match(r"^assign\s+(\$\w+)\s*=\s*(\$\w+)$", key)
    if m:
        return AssignClause(
            target_bind_var=m.group(1),
            source_bind_var=m.group(2),
            cast_type_pattern=val,
        )

    # mock: ...
    if key == "mock":
        return _parse_mock_clause(val)

    raise ValueError(f"Unknown given clause key: '{key}'")


def _parse_create_clause(bind_var: str, value: str) -> CreateClause:
    m = re.match(r"^([\w<>,\s?]+?)(?:\(([^)]*)\))?$", value)
    if not m:
        raise ValueError(f"Cannot parse create value: '{value}'")

    clause = CreateClause(bind_var=bind_var, type_pattern=m.group(1).strip())
    if m.group(2) is not None:
        clause.exact_arg_count = True
        if m.group(2).strip():
            clause.arg_bindings = _split_args(m.group(2))
    return clause


def _parse_call_clause(value: str, result_bind_var: str | None = None) -> CallClause:
    m = re.match(
        r"^(?:(\$\w+)\.)?(\w+)(?:\(([^)]*)\))?\s*(?:on\s+(.+?))?\s*(?:->\s*(\$\w+))?\s*(?:throws\s+(\$\w+|\w+))?$",
        value,
    )
    if not m:
        raise ValueError(f"Cannot parse call value: '{value}'")

    clause = CallClause(result_bind_var=result_bind_var, method_name=m.group(2))
    if m.group(1):
        clause.receiver_bind_var = m.group(1)
    if m.group(3) and m.group(3).strip():
        clause.arg_bindings = _split_args(m.group(3))
    if m.group(4):
        clause.on_type_pattern = m.group(4).strip()
    if m.group(5):
        clause.callback_bind_var = m.group(5)
    if m.group(6):
        clause.throws_bind_var = m.group(6)
    return clause


def _parse_mock_clause(value: str) -> MockClause:
    m = re.match(r"^(?:(\$\w+)\.)?(\w+)\s+returns\s+(.+)$", value)
    if not m:
        raise ValueError(f"Cannot parse mock value: '{value}'")
    return MockClause(
        mock_bind_var=m.group(1) if m.group(1) else None,
        method_name=m.group(2),
        return_value=m.group(3).strip(),
    )


# ── Assert clause parsing ────────────────────────────

def _parse_assert_clause(mapping: dict) -> AssertClause:
    key, value = next(iter(mapping.items()))
    clause = AssertClause(semantic_meaning=key)

    if isinstance(value, list):
        clause.args = [_normalize_arg(o) for o in value]
    elif value is not None:
        s = _normalize_arg(value)
        if s != "any":
            clause.args.append(s)

    return clause


def _normalize_arg(o) -> str:
    if o is None:
        return "None"
    s = str(o)
    if len(s) == 0:
        return '""'
    return s


def _split_args(arg_str: str) -> list[str]:
    return [a.strip() for a in arg_str.split(",") if a.strip()]

File: CSharp/Services/RuleParser.cs
using System.Text.RegularExpressions;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

/// <summary>
/// Parses YAML rule files into Rule objects.
///
/// YAML format — each given/assert item is a single-key YAML mapping:
///
///   rules:
///     - id: my_rule
///       description: "..."
///       given:
///         - create $acc: Accessor($val)
///         - call: $setter.SetValue($newVal) on IAccessorSetter&lt;any&gt;
///         - assign $setter = $acc: IAccessorSetter&lt;any&gt;
///         - mock: OtherFunc returns false
///       assert:
///         - IsNull: $acc.Value
///         - AreEqual: [$a, $b]
///         - Throws: any
/// </summary>
static class RuleParser
{
    public static List<Rule> Parse(string yamlContent)
    {
        var deserializer = new DeserializerBuilder()
            .WithNamingConvention(UnderscoredNamingConvention.Instance)
            .Build();

        var doc = deserializer.Deserialize<YamlRoot>(yamlContent);

        return doc.Rules.Select(ParseRule).ToList();
    }

    static Rule ParseRule(YamlRule yr)
    {
        var rule = new Rule { Id = yr.Id, Parent = yr.Parent, OrGroup = yr.OrGroup, Description = yr.Description ?? "" };

        foreach (var givenMap in yr.Given ?? new())
            rule.Given.Add(ParseGivenClause(givenMap));

        foreach (var assertMap in yr.Assert ?? new())
            rule.Assert.Add(ParseAssertClause(assertMap));

        return rule;
    }

    // ── Given clause parsing ─────────────────────────────

    static GivenClause ParseGivenClause(Dictionary<string, object> map)
    {
        var (key, value) = map.First();
        var val = value?.ToString() ?? "";

        // Check for "not " prefix → negated clause
        bool negated = false;
        if (key.StartsWith("not ", StringComparison.OrdinalIgnoreCase))
        {
            negated = true;
            key = key[4..]; // strip "not "
        }

        GivenClause clause = ParseGivenClauseCore(key, val);
        clause.Negated = negated;
        return clause;
    }

    static GivenClause ParseGivenClauseCore(string key, string val)
    {
        // create $var: TypePattern($arg1, $arg2)
        // YAML parses "create $acc" as key, "Accessor($val)" as value
        var createMatch = Regex.Match(key, @"^create\s+(\$\w+)$");
        if (createMatch.Success)
            return ParseCreateClause(createMatch.Groups[1].Value, val);

        // $result = call: ...
        var resultCallMatch = Regex.Match(key, @"^(\$\w+)\s*=\s*call$");
        if (resultCallMatch.Success)
            return ParseCallClause(val, resultBindVar: resultCallMatch.Groups[1].Value);

        // call: ...
        if (key == "call")
            return ParseCallClause(val);

        // assign $target = $source: CastType
        var assignMatch = Regex.Match(key, @"^assign\s+(\$\w+)\s*=\s*(\$\w+)$");
        if (assignMatch.Success)
        {
            return new AssignClause
            {
                TargetBindVar = assignMatch.Groups[1].Value,
                SourceBindVar = assignMatch.Groups[2].Value,
                CastTypePattern = val
            };
        }

        // mock: ...
        if (key == "mock")
            return ParseMockClause(val);

        throw new InvalidOperationException($"Unknown given clause key: '{key}'");
    }

    static CreateClause ParseCreateClause(string bindVar, string value)
    {
        // value: "Accessor" or "Accessor<any>" or "Accessor($val, $arg2)"
        var match = Regex.Match(value, @"^([\w<>,\s?]+?)(?:\(([^)]*)\))?$");
        if (!match.Success)
            throw new InvalidOperationException($"Cannot parse create value: '{value}'");

        var clause = new CreateClause
        {
            BindVar = bindVar,
            TypePattern = match.Groups[1].Value.Trim()
        };

        if (match.Groups[2].Success)
        {
            clause.ExactArgCount = true;
            if (!string.IsNullOrWhiteSpace(match.Groups[2].Value))
                clause.ArgBindings = SplitArgs(match.Groups[2].Value);
        }

        return clause;
    }

    static CallClause ParseCallClause(string value, string? resultBindVar = null)
    {
        // value: "MethodName" or "$recv.Method($arg)" or "Method on Type" or "Method($a) -> $cb" or "Method throws $exc"
        // Full pattern: [receiver.]Method[(args)] [on TypePattern] [-> $cbVar] [throws $excVar]
        var match = Regex.Match(value, @"^(?:(\$\w+)\.)?(\w+)(?:\(([^)]*)\))?\s*(?:on\s+(.+?))?\s*(?:->\s*(\$\w+))?\s*(?:throws\s+(\$\w+|\w+))?$");
        if (!match.Success)
            throw new InvalidOperationException($"Cannot parse call value: '{value}'");

        var clause = new CallClause
        {
            ResultBindVar = resultBindVar,
            MethodName = match.Groups[2].Value
        };

        if (match.Groups[1].Success && !string.IsNullOrEmpty(match.Groups[1].Value))
            clause.ReceiverBindVar = match.Groups[1].Value;

        if (match.Groups[3].Success && !string.IsNullOrWhiteSpace(match.Groups[3].Value))
            clause.ArgBindings = SplitArgs(match.Groups[3].Value);

        if (match.Groups[4].Success && !string.IsNullOrEmpty(match.Groups[4].Value))
            clause.OnTypePattern = match.Groups[4].Value.Trim();

        if (match.Groups[5].Success && !string.IsNullOrEmpty(match.Groups[5].Value))
            clause.CallbackBindVar = match.Groups[5].Value;

        if (match.Groups[6].Success && !string.IsNullOrEmpty(match.Groups[6].Value))
            clause.ThrowsBindVar = match.Groups[6].Value;

        return clause;
    }

    static MockClause ParseMockClause(string value)
    {
        // value: "OtherFunc returns false" or "$mock.OtherFunc returns false"
        var match = Regex.Match(value, @"^(?:(\$\w+)\.)?(\w+)\s+returns\s+(.+)$");
        if (!match.Success)
            throw new InvalidOperationException($"Cannot parse mock value: '{value}'");

        return new MockClause
        {
            MockBindVar = match.Groups[1].Success ? match.Groups[1].Value : null,
            MethodName = match.Groups[2].Value,
            ReturnValue = match.Groups[3].Value.Trim()
        };
    }

    // ── Assert clause parsing ────────────────────────────

    static AssertClause ParseAssertClause(Dictionary<string, object> map)
    {
        var (key, value) = map.First();
        var clause = new AssertClause { SemanticMeaning = key };

        if (value is List<object> list)
        {
            clause.Args = list.Select(o =>
            {
                if (o == null) return "null";
                var s = o.ToString() ?? "null";
                // YAML "" is an empty string; C# AST represents it as ""
                if (s.Length == 0) return "\"\"";
                return s;
            }).ToList();
        }
        else if (value != null)
        {
            var s = value.ToString() ?? "";
            if (s.Length == 0) s = "\"\"";
            if (s != "any")
                clause.Args.Add(s);
        }

        return clause;
    }

    // ── Utility ──────────────────────────────────────────

    static List<string> SplitArgs(string argStr)
    {
        return argStr.Split(',').Select(a => a.Trim()).Where(a => a.Length > 0).ToList();
    }

    // ── YAML deserialization models ──────────────────────

    class YamlRoot
    {
        public List<YamlRule> Rules { get; set; } = new();
    }

    class YamlRule
    {
        public string Id { get; set; } = "";
        public string? Parent { get; set; }
        public string? OrGroup { get; set; }
        public string? Description { get; set; }
        public List<Dictionary<string, object>>? Given { get; set; }
        public List<Dictionary<string, object>>? Assert { get; set; }
    }
}
